limitador/src/storage/redis/counters_cache.rs

```diff
@@ -20,46 +22,91 @@ pub struct CachedCounterValue {
 }
 
 pub struct Batcher {
-    updates: Mutex<HashMap<Counter, Arc<CachedCounterValue>>>,
+    updates: DashMap<Counter, Arc<CachedCounterValue>>,
     notifier: Notify,
     interval: Duration,
+    priority_flush: AtomicBool,
 }
 
 impl Batcher {
     fn new(period: Duration) -> Self {
         Self {
-            updates: Mutex::new(Default::default()),
+            updates: Default::default(),
             notifier: Default::default(),
             interval: period,
+            priority_flush: AtomicBool::new(false),
         }
     }
 
     pub fn is_empty(&self) -> bool {
-        self.updates.lock().unwrap().is_empty()
+        self.updates.is_empty()
     }
 
-    pub async fn consume(&self, min: usize) -> HashMap<Counter, Arc<CachedCounterValue>> {
+    pub async fn consume<F, Fut, O>(&self, min: usize, consumer: F) -> O
+    where
+        F: FnOnce(HashMap<Counter, Arc<CachedCounterValue>>) -> Fut,
+        Fut: Future<Output = O>,
+    {
         let mut interval = interval(self.interval);
-        let mut ready = self.updates.lock().unwrap().len() >= min;
+        let mut ready = self.updates.len() >= min;
         loop {
             if ready {
-                return self.consume_all();
+                let mut batch = Vec::with_capacity(min);
+                let mut probably_fake = Vec::with_capacity(min);
+                for entry in &self.updates {
+                    if entry.value().value.ttl() < self.interval {
+                        batch.push(entry.key().clone());
+                        if batch.len() == min {
+                            break;
+                        }
+                    }
+                    if entry.value().expiry.duration() == Duration::from_secs(entry.key().seconds())
+                    {
+                        probably_fake.push(entry.key().clone());
+                        if probably_fake.len() == min {
+                            break;
+                        }
+                    }
+                }
+                if let Some(remaining) = min.checked_sub(batch.len()) {
+                    let take = probably_fake.into_iter().take(remaining);
+                    batch.append(&mut take.collect());
+                }
+                if let Some(remaining) = min.checked_sub(batch.len()) {
+                    let take = self.updates.iter().take(remaining);
+                    batch.append(&mut take.map(|e| e.key().clone()).collect());
+                }
+                let mut result = HashMap::new();
+                for counter in &batch {
+                    let value = self.updates.get(counter).unwrap().clone();
+                    result.insert(counter.clone(), value);
+                }
+                let result = consumer(result).await;
+                for counter in &batch {
+                    self.updates
+                        .remove_if(counter, |_, v| v.no_pending_writes());
+                }
```

Comment:
If _after_ having flushed to redis no additional pending writes were added, we can remove these entries from the queue.